### ISSUE TITLE
`metadata.bam_index ` is None if `argument` and `name` are used in a tool input

### ISSUE THREAD

**Issue by mvdbeek** [galaxyproject/galaxy#3930]
for example
```
<param name="input" argument="--input_path" type="data" format="bam"/>
```
and referencing the bam index with `ln -f -s $input.metadata.bam_index input.bam.bai` fails like this:
```
...
ln -f -s None input.bam.bai &&
...
```

Taking out `argument="--input_path"` restores that functionality.

**Comment by mvdbeek** [comment]
This is not a big deal, but unexpected and potentially a time-sink.

**Comment by nsoranzo** [comment]
Uh, it would be good to bisect this to apply the fix to the first affected release.

**Comment by jmchilton** [comment]
Using this diff:

```
diff --git a/test/functional/tools/metadata_bam.xml b/test/functional/tools/metadata_bam.xml
index 2f20be5b53..beb913acce 100644
--- a/test/functional/tools/metadata_bam.xml
+++ b/test/functional/tools/metadata_bam.xml
@@ -1,10 +1,10 @@
 <tool id="metadata_bam" name="metadata BAM" version="1.0.0">
-  <command>echo "${ref_names}" &gt; "${output_of_input_metadata}"</command>
+  <command>echo "${ref_names}" &gt; "${output_of_input_metadata}"; echo $input.metadata.bam_index</command>
   <inputs>
-    <param name="input_bam" type="data" format="bam" label="BAM File"/>
+    <param name="input" type="data" format="bam" label="BAM File" argument="--cow" />
     <param name="ref_names" type="select" optional="False" label="Select references you would like to restrict bam to" multiple="True">
         <options>
-            <filter type="data_meta" ref="input_bam" key="reference_names" />
+            <filter type="data_meta" ref="input" key="reference_names" />
         </options>
     </param>
   </inputs>
@@ -13,7 +13,7 @@
   </outputs>
   <tests>
     <test>
-      <param name="input_bam" value="3.bam" ftype="bam" />
+      <param name="input" value="3.bam" ftype="bam" />
       <!-- This output tests setting using bam reference name metadata above -->
       <param name="ref_names" value="chr10_random,chr11,chrM,chrX,chr16" />
       <output name="output_of_input_metadata" ftype="txt">
```

applied the metadata_bam test tool I failed to reproduce this error in dev or on release_17.01.

Did I mess up, did you mess up, or do we have something more subtle broken like the dataset being ready in some cases before the metadata has been written leading to the tool transiently failing based on timing errors - that made it seem like ``argument`` may have been the difference. Did you always test on green box and get this error or was the testing part of a workflow for instance?

**Comment by mvdbeek** [comment]
It was part of a workflow initiailly, I'll test it with your diff.

**Comment by mvdbeek** [comment]
Hmm, I can't reproduce this anymore, not with the test-tool nor with the initial tool that I had noticed this behavior with. I guess I was misled because the re-run with the changed parameter worked. I'll keep an eye on this, but maybe this was indeed related to a workflow problem (I noticed that sometimes my datasets turn green, but the metadata isn't being updated for a minute or two ... perhaps that's related). I will open this again if I can figure out more.
